pkg/maps/ctmap/ctmap.go

```diff
@@ -510,6 +517,102 @@ func GC(m *Map, filter *GCFilter) int {
 	return doGC(m, filter)
 }
 
+// PurgeOrphanNATEntries removes orphan SNAT entries. We call an SNAT entry
+// orphan if it does not have either a corresponding CT entry or an SNAT entry
+// in reverse order.
+//
+// The former case can happen when the CT entry is removed by the LRU eviction
+// which happens when the CT map becomes full. The latter case can happen for
+// the same LRU eviction reasons.
+//
+// PurgeOrphanNATEntries() is triggered by the datapath via the GC signaling
+// mechanism. When the datapath SNAT fails to find free mapping after
+// SNAT_SIGNAL_THRES attempts, it sends the signal via the perf ring buffer.
+// The consumer of the buffer invokes the function.
+//
+// The SNAT is being used for the following cases:
+// 1. By NodePort BPF on an intermediate node before fwd'ing request from outside
+//     to a destination node.
+// 2. A packet from local endpoint sent to outside (BPF-masq).
+// 3. A packet from a host local application (i.e. running in the host netns)
+//    This is needed to prevent SNAT from hijacking such connections.
+// 4. By DSR on a backend node to SNAT responses with service IP+port before
+//    sending to a client.
+//
+// In the case of 1-3, we always create a CT_EGRESS CT entry. This allows the
+// CT GC to remove corresponding SNAT entries. In the case of 4, will create
+// CT_INGRESS CT entry. See the unit test TestOrphanNatGC for more examples.
+//
+// The function only handles 1-3 cases, the 4. case is TODO(brb).
+func PurgeOrphanNATEntries(ctMap *Map) *NatGCStats {
+	if option.Config.NodePortMode == option.NodePortModeDSR ||
+		option.Config.NodePortMode == option.NodePortModeHybrid {
+		return nil
+	}
+
+	natMap := mapInfo[ctMap.mapType].natMap
+	if natMap == nil {
+		return nil
+	}
+	stats := newNatGCStats(natMap)
+
+	cb := func(key bpf.MapKey, value bpf.MapValue) {
+		var ctKey tuple.TupleKey4Global
+		natKey := key.(*nat.NatKey4)
+		natVal := value.(*nat.NatEntry4)
+
+		if natKey.GetFlags()&tuple.TUPLE_F_IN == 0 { // dir == CT_EGRESS
+			ctKey.Flags = tuple.TUPLE_F_OUT
+			// Workaround #5848
+			ctKey.SourceAddr = natKey.DestAddr
+			ctKey.DestAddr = natKey.SourceAddr
+			ctKey.SourcePort = natKey.SourcePort
+			ctKey.DestPort = natKey.DestPort
+			ctKey.NextHeader = natKey.NextHeader
+			// TODO(brb) Check for CT entry regardless of direction
+			if _, err := ctMap.Lookup(&ctKey); err != nil && errors.Is(err, unix.ENOENT) {
```

Comment:
I don't quite follow why we do this SNAT map GC here in addition to the regular one? That seems a _very_ high CPU cost to pay. When we get the signal from BPF data path that there is pressure in finding a new SNAT mapping in the NAT table then idea was to walk all NAT entries, find orphans but given you already do the CT lookup here anyway, we already have the related CT entry at hand too, so it's low-cost to delete it from right here. That means instead of walking CT + SNAT map, we only walk SNAT map and delete CT from there. Is there any rationale that this is not feasible?

Reply:
> Is there any rationale that this is not feasible?

Maybe I'm missing something, but I don't see how it's possible, because of CT_SERVICE entries and also, not all CT entries have related NAT entries (e.g. pod to pod).

Reply:
Right, I mean technically, they would be evicted via LRU or regular time-based GC. The CT_SERVICE ones are more annoying though given we want to improve triggering new backend selection. Technically it's an entirely different issue and cleanups could be triggered on backend updates, but the regular CT walk as you have in addition in here would help it too, just the cost becomes huge.